shared/js/moz_intl.js

```diff
@@ -0,0 +1,199 @@
+(function(global) {
+'use strict';
+
+global.mozIntl = {
+  /**
+   * Format an Array of strings using locale specific separators.
+   * Currently it only uses middle separator resulting in a local
+   * equivalent of 'X, Y, Z'.
+   *
+   * In the Intl API it will support start/end separators allowing for
+   * things like 'X, Y and Z'.
+   *
+   * @param {Array} An array of strings to be formatted
+   * @returns {Promise} A promise of a string
+   */
+  formatList: function(list) {
+    return navigator.mozL10n.formatValue('listSeparator_middle').then(sep => {
+      return list.join(sep);
+    });
+  },
+
+  /**
+   * Overload of Intl.DateTimeFormat that extends with future features.
+   *
+   * 1) dayperiod token
+   *
+   * `dayperiod` token allows to decide whether the formatted time should
+   * contain dayperiod token. The value may be true or false.
+   *
+   * Example:
+   *
+   * mozIntl.DateTimeFormat(navigator.languages, {
+   *   dayperiod: true,
+   *   hour12: navigator.mozHour12,
+   *   hour: 'numeric',
+   *   minute: 'numeric'
+   * });
+   *
+   * 2) Token formatting
+   *
+   * This feature allows for formatting of tokens in the result string.
+   *
+   * Example:
+   *
+   * var f = mozIntl.DateTimeFormat(navigator.languages, {
+   *   weekday: 'long',
+   *   day: 'numeric',
+   *   month: 'long'
+   * });
+   * 
+   * f.format(date, {
+   *  day: '<strong>$&</strong>',
+   * });
+   */
+  DateTimeFormat: function(locales, options) {
+    const resolvedOptions = Object.assign({}, options);
+
+    if (options.dayperiod) {
+      if (!('hour' in options)) {
+        resolvedOptions.hour = 'numeric';
+      }
+    }
+
+    if (resolvedOptions.hour) {
+      resolvedOptions.hour12 = navigator.mozHour12;
+    }
+
+    var intlFormat = new Intl.DateTimeFormat(locales, resolvedOptions);
+
+    return {
+      format: function(date, tokenFormats) {
+        var dayPeriod;
+
+        var string = intlFormat.format(date);
+
+        if (options.dayperiod === false &&
+            intlFormat.resolvedOptions().hour12 === true) {
+          dayPeriod = date.toLocaleFormat('%p');
+          string = string.replace(dayPeriod, '').trim();
+        } else if (options.dayperiod === true &&
+           options.hour === undefined) {
+          dayPeriod = date.toLocaleFormat('%p');
+          const hour = date.toLocaleString(navigator.languages, {
+            hour12: true,
+            hour: 'numeric'
+          }).replace(dayPeriod, '').trim();
+          string = string.replace(hour, '').trim();
+        }
+
+        for (var token in tokenFormats) {
+          if (token === 'dayperiod' &&
+            intlFormat.resolvedOptions().hour12 === false) {
+            continue;
+          }
+          const localOptions = {};
+          localOptions[token] = resolvedOptions[token];
+
+          var formatter = global.mozIntl.DateTimeFormat(
+            navigator.languages, localOptions);
```

Comment:
per formatting token, yes. I need to isolate the token I'm going to run a replace on, so I need to retrieve it.